python/cuml/test/test_pickle.py

```diff
@@ -0,0 +1,283 @@
+        y_cudf = make_cudf_series(y_train)
+        X_cudf_test = np_to_cudf(X_test)
+
+        cu_before_pickle_model.fit(X_cudf, y_cudf)
+        cu_before_pickle_predict = cu_before_pickle_model.predict(X_cudf_test).to_array()
+
+    else:
+        cu_before_pickle_model.fit(X_train, y_train)
+        cu_before_pickle_predict = cu_before_pickle_model.predict(X_test).to_array()
+
+    cu_after_pickle_model = pickle_save_load(serializer_type, cu_before_pickle_model)
+
+    if input_type == 'dataframe':
+        cu_after_pickle_predict = cu_after_pickle_model.predict(X_cudf_test).to_array()
+    else:
+        cu_after_pickle_predict = cu_after_pickle_model.predict(X_test).to_array()
+
+    assert array_equal(cu_before_pickle_predict, cu_after_pickle_predict, with_sign=True)
+
+@pytest.mark.parametrize('datatype', [np.float32])
+@pytest.mark.parametrize('input_type', ['dataframe'])
+@pytest.mark.parametrize('model', ['KMeans'])
+@pytest.mark.parametrize('model_objects', [cluster_models])
+@pytest.mark.parametrize('nrows', [unit_param(20)])
+@pytest.mark.parametrize('ncols', [unit_param(3)])
+@pytest.mark.parametrize('serializer_type', ['pickle', 'dill'])
+@pytest.mark.xfail
+def test_cluster_pickle(datatype, input_type, model, model_objects, nrows, ncols,
+                        serializer_type):
+    train_rows = np.int32(nrows*0.8)
+    X, _ = make_blobs(n_samples=(nrows), n_features=ncols, random_state=0)
+    X_test = np.asarray(X[train_rows:, :]).astype(datatype)
+    X_train = np.asarray(X[:train_rows, :]).astype(datatype)
+
+    cu_before_pickle_model = model_objects[model]
+
+    if input_type == 'dataframe':
+        X_cudf = np_to_cudf(X_train)
+        X_cudf_test = np_to_cudf(X_test)
+
+        cu_before_pickle_model.fit(X_cudf)
+        cu_before_pickle_predict = cu_before_pickle_model.predict(X_cudf_test).to_array()
+
+    else:
+        cu_before_pickle_model.fit(X_train)
+        cu_before_pickle_predict = cu_before_pickle_model.predict(X_test).to_array()
+
+    cu_after_pickle_model = pickle_save_load(serializer_type, cu_before_pickle_model)
+
+    if input_type == 'dataframe':
+        cu_after_pickle_predict = cu_after_pickle_model.predict(X_cudf_test).to_array()
+    else:
+        cu_after_pickle_predict = cu_after_pickle_model.predict(X_test).to_array()
+
+    assert array_equal(cu_before_pickle_predict, cu_after_pickle_predict, with_sign=True)
+
+@pytest.mark.parametrize('datatype', [np.float32])
+@pytest.mark.parametrize('input_type', ['dataframe'])
+@pytest.mark.parametrize('model', ['PCA', 'TruncatedSVD', 'UMAP'])
+@pytest.mark.parametrize('model_objects', [decomposition_models])
+@pytest.mark.parametrize('nrows', [unit_param(20)])
+@pytest.mark.parametrize('ncols', [unit_param(3)])
+@pytest.mark.parametrize('serializer_type', ['pickle', 'dill'])
+@pytest.mark.xfail
+def test_decomposition_pickle(datatype, input_type, model, model_objects, nrows, ncols,
+                        serializer_type):

```

Comment:
These various tests seem very similar (decomp, cluster, reg, etc.). Can they call an underlying function that does most of the heavy lifting except for the one bit of difference? (Sorry, I know I was suggesting they could be separate functions to have their own params, etc., but I was thinking about sort of an underlying utility function)

Reply:
I have created a utility function and you can see it in my latest commits. Let me know if this looks cleaner